File: cb/api/xml_builder.py
from __future__ import annotations
"""Config XML builders for Jobs, Nodes, Credentials -- using stdlib xml.etree."""

import xml.etree.ElementTree as ET


def _xml_str(root: ET.Element) -> str:
    """Serialise an ElementTree to a UTF-8 XML string with declaration."""
    ET.indent(root, space="  ")
    return (
        "<?xml version='1.1' encoding='UTF-8'?>\n"
        + ET.tostring(root, encoding="unicode")
    )


# -- Job XML ---------------------------------------------------


def inject_email_publisher(publishers: ET.Element, email: str, email_cond: str):
    """Helper to inject ExtendedEmailPublisher into publishers block."""
    ext_mail = ET.SubElement(publishers, "hudson.plugins.emailext.ExtendedEmailPublisher", {"plugin": "email-ext"})
    ET.SubElement(ext_mail, "recipientList").text = email
    ET.SubElement(ext_mail, "configuredTriggers")
    cur_triggers = ext_mail.find("configuredTriggers")
    
    def _add_email_trigger(parent, tag_name):
        evt = ET.SubElement(parent, tag_name)
        eml = ET.SubElement(evt, "email")
        ET.SubElement(eml, "subject").text = "$PROJECT_DEFAULT_SUBJECT"
        ET.SubElement(eml, "body").text = "$PROJECT_DEFAULT_CONTENT"
        ET.SubElement(eml, "recipientList").text = ""
        rp = ET.SubElement(eml, "recipientProviders")
        ET.SubElement(rp, "hudson.plugins.emailext.plugins.recipients.ListRecipientProvider")
        ET.SubElement(eml, "attachmentsPattern").text = ""
        ET.SubElement(eml, "attachBuildLog").text = "false"
        ET.SubElement(eml, "compressBuildLog").text = "false"
        ET.SubElement(eml, "replyTo").text = "$PROJECT_DEFAULT_REPLYTO"
        ET.SubElement(eml, "contentType").text = "project"

    if email_cond in ("failed", "always"):
        _add_email_trigger(cur_triggers, "hudson.plugins.emailext.plugins.trigger.FailureTrigger")
    if email_cond in ("success", "always"):
        _add_email_trigger(cur_triggers, "hudson.plugins.emailext.plugins.trigger.SuccessTrigger")
        
    ET.SubElement(ext_mail, "contentType").text = "default"
    ET.SubElement(ext_mail, "defaultSubject").text = "$DEFAULT_SUBJECT"
    ET.SubElement(ext_mail, "defaultContent").text = "$DEFAULT_CONTENT"
    ET.SubElement(ext_mail, "attachmentsPattern").text = ""
    ET.SubElement(ext_mail, "presendScript").text = "$DEFAULT_PRESEND_SCRIPT"
    ET.SubElement(ext_mail, "postsendScript").text = "$DEFAULT_POSTSEND_SCRIPT"
    ET.SubElement(ext_mail, "attachBuildLog").text = "false"
    ET.SubElement(ext_mail, "compressBuildLog").text = "false"
    ET.SubElement(ext_mail, "replyTo").text = "$DEFAULT_REPLYTO"
    ET.SubElement(ext_mail, "from").text = ""
    ET.SubElement(ext_mail, "saveOutput").text = "false"
    ET.SubElement(ext_mail, "disabled").text = "false"

def build_freestyle_xml(
        desc: str = "", 
        shell_cmd: str = "echo hello", 
        node: str | None = None,
        chdir: str | None = None,
        schedule: str | None = None,
        email: str | None = None,
        email_cond: str = "failed"
) -> str:
    """Freestyle project config.xml."""
    root = ET.Element("project")
    ET.SubElement(root, "description").text = desc
    ET.SubElement(root, "keepDependencies").text = "false"
    ET.SubElement(root, "properties")
    ET.SubElement(root, "scm", {"class": "hudson.scm.NullSCM"})
    ET.SubElement(root, "canRoam").text = "false" if node else "true"
    if node:
        ET.SubElement(root, "assignedNode").text = node
    ET.SubElement(root, "disabled").text = "false"
    
    triggers = ET.SubElement(root, "triggers")
    if schedule:
        timer = ET.SubElement(triggers, "hudson.triggers.TimerTrigger")
        ET.SubElement(timer, "spec").text = schedule
        
    builders = ET.SubElement(root, "builders")
    shell = ET.SubElement(builders, "hudson.tasks.Shell")
    final_cmd = f"cd {chdir} && {shell_cmd}" if chdir else shell_cmd
    ET.SubElement(shell, "command").text = final_cmd
    
    publishers = ET.SubElement(root, "publishers")
    if email:
        inject_email_publisher(publishers, email, email_cond)

    ET.SubElement(root, "buildWrappers")
    return _xml_str(root)



def build_pipeline_xml(
        desc: str = "", 
        script: str = "pipeline { agent any\n  stages { stage('Build') { steps { echo 'Hello' } } } }", 
        node: str | None = None,
        schedule: str | None = None
) -> str:
    """Pipeline (WorkflowJob) config.xml."""
    root = ET.Element(
        "flow-definition",
        {"plugin": "workflow-job"}
    )
    ET.SubElement(root, "description").text = desc

    if node:
        ET.SubElement(root, "assignedNode").text = node

    actions = ET.SubElement(root, "actions")
    defn = ET.SubElement(
        root, "definition",
        {"class": "org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition",
         "plugin": "workflow-cps"}
    )
    ET.SubElement(defn, "script").text = script
    ET.SubElement(defn, "sandbox").text = "true"
    
    triggers = ET.SubElement(root, "triggers")
    if schedule:
        timer = ET.SubElement(triggers, "hudson.triggers.TimerTrigger")
        ET.SubElement(timer, "spec").text = schedule
        
    ET.SubElement(root, "disabled").text = "false"
    return _xml_str(root)


def build_folder_xml(desc: str = "") -> str:
    """CloudBees Folder config.xml."""
    root = ET.Element(
        "com.cloudbees.hudson.plugins.folder.Folder",
        {"plugin": "cloudbees-folder"}
    )
    ET.SubElement(root, "description").text = desc
    ET.SubElement(root, "views")
    ET.SubElement(root, "primaryView").text = "All"
    ET.SubElement(root, "healthMetrics")
    return _xml_str(root)


# -- Node XML --------------------------------------------------

def build_permanent_node_xml(
    name: str,
    remote_dir: str,
    num_executors: int = 1,
    labels: str = "",
    desc: str = "",
    host: str = "",
    port: int = 22,
    credentials_id: str = "",
) -> str:
    """Config XML for Permanent Agent (SSH or JNLP)."""
    root = ET.Element("slave")
    ET.SubElement(root, "name").text = name
    ET.SubElement(root, "description").text = desc
    ET.SubElement(root, "remoteFS").text = remote_dir
    ET.SubElement(root, "numExecutors").text = str(num_executors)
    ET.SubElement(root, "mode").text = "NORMAL"
    ET.SubElement(root, "retentionStrategy", {"class": "hudson.slaves.RetentionStrategy$Always"})
    
    if host:
        launcher = ET.SubElement(root, "launcher", {"class": "hudson.plugins.sshslaves.SSHLauncher", "plugin": "ssh-slaves"})
        ET.SubElement(launcher, "host").text = host
        ET.SubElement(launcher, "port").text = str(port)
        ET.SubElement(launcher, "credentialsId").text = credentials_id
        ET.SubElement(launcher, "sshHostKeyVerificationStrategy", {"class": "hudson.plugins.sshslaves.verifiers.NonVerifyingKeyVerificationStrategy"})
    else:
        launcher = ET.SubElement(root, "launcher", {"class": "hudson.slaves.JNLPLauncher"})
        wds = ET.SubElement(launcher, "workDirSettings")
        ET.SubElement(wds, "disabled").text = "false"
        ET.SubElement(wds, "internalDir").text = "remoting"
        ET.SubElement(wds, "failIfWorkDirIsMissing").text = "false"
        
    ET.SubElement(root, "label").text = labels
    ET.SubElement(root, "nodeProperties")
    return _xml_str(root)


# -- Credential XML --------------------------------------------


def build_username_password_cred_xml(
    cred_id: str,
    username: str,
    password: str,
    desc: str = "",
    scope: str = "GLOBAL",
) -> str:
    """UsernamePassword credential config.xml."""
    root = ET.Element(
        "com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl"
    )
    ET.SubElement(root, "scope").text = scope
    ET.SubElement(root, "id").text = cred_id
    ET.SubElement(root, "description").text = desc
    ET.SubElement(root, "username").text = username
    ET.SubElement(root, "password").text = password
    return _xml_str(root)

File: src/plugins/job/xml-builder.ts
/**
 * Freestyle/Folder job config.xml builders — the job-specific XML.
 *
 * Hand-builds indented XML (2-space) with the prolog
 * `<?xml version='1.1' encoding='UTF-8'?>` to match Python ET.indent() output.
 *
 * Email-ext and TimerTrigger XML are NOT here — they live in domain/email and
 * domain/schedule (shared, leaf, unit-tested). This file composes them into a
 * `<project>`.
 */

import { escapeXml, xmlParser } from "../../domain/xml";
import { buildEmailPublisherBlock } from "../../domain/email";
import { buildTimerTriggerBlock } from "../../domain/schedule";
import type { StringParamDef, PipelineXmlOpts } from "./types";

/**
 * Build the `<properties>` block for a job. When `params` is empty, returns the
 * self-closing `<properties/>` (matches the legacy default). Otherwise emits a
 * `ParametersDefinitionProperty` holding one `StringParameterDefinition` per
 * param. `indent` is the leading whitespace for the `<properties>` tag.
 */
export function buildParametersProperty(
  params: StringParamDef[] | null | undefined,
  indent = "  ",
): string {
  if (!params || params.length === 0) return `${indent}<properties/>`;

  const i2 = `${indent}  `;
  const i3 = `${indent}    `;
  const i4 = `${indent}      `;

  const lines: string[] = [
    `${indent}<properties>`,
    `${i2}<hudson.model.ParametersDefinitionProperty>`,
    `${i3}<parameterDefinitions>`,
  ];

  for (const p of params) {
    lines.push(`${i4}<hudson.model.StringParameterDefinition>`);
    lines.push(`${i4}  <name>${escapeXml(p.name)}</name>`);
    lines.push(`${i4}  <description>${escapeXml(p.description ?? "")}</description>`);
    lines.push(`${i4}  <defaultValue>${escapeXml(p.defaultValue ?? "")}</defaultValue>`);
    lines.push(`${i4}  <trim>false</trim>`);
    lines.push(`${i4}</hudson.model.StringParameterDefinition>`);
  }

  lines.push(`${i3}</parameterDefinitions>`);
  lines.push(`${i2}</hudson.model.ParametersDefinitionProperty>`);
  lines.push(`${indent}</properties>`);
  return lines.join("\n");
}

/**
 * Build a Freestyle project config.xml.
 * Mirrors Python build_freestyle_xml() exactly.
 */
export function buildFreestyleXml(opts: {
  desc?: string;
  shellCmd?: string;
  node?: string | null;
  chdir?: string | null;
  schedule?: string | null;
  email?: string | null;
  emailCond?: string;
  emailKeywords?: string[] | null;
  emailRegex?: string | null;
  params?: StringParamDef[] | null;
  concurrent?: boolean;
}): string {
  const {
    desc = "",
    shellCmd = "echo hello",
    node = null,
    chdir = null,
    schedule = null,
    email = null,
    emailCond = "failed",
    emailKeywords = null,
    emailRegex = null,
    params = null,
    concurrent = false,
  } = opts;

  // Quote chdir so paths with spaces survive the shell (e.g. "cd /my dir && cmd").
  const finalCmd = chdir ? `cd "${chdir}" && ${shellCmd}` : shellCmd;
  const canRoam = node ? "false" : "true";

  const lines: string[] = [
    "<?xml version='1.1' encoding='UTF-8'?>",
    "<project>",
    `  <description>${escapeXml(desc)}</description>`,
    "  <keepDependencies>false</keepDependencies>",
    buildParametersProperty(params, "  "),
    '  <scm class="hudson.scm.NullSCM"/>',
    `  <canRoam>${canRoam}</canRoam>`,
  ];

  if (node) {
    lines.push(`  <assignedNode>${escapeXml(node)}</assignedNode>`);
  }

  lines.push("  <disabled>false</disabled>");
  lines.push("  <triggers>");

  const timer = buildTimerTriggerBlock(schedule, "    ");
  if (timer) lines.push(timer);

  lines.push("  </triggers>");
  lines.push(`  <concurrentBuild>${concurrent ? "true" : "false"}</concurrentBuild>`);
  lines.push("  <builders>");
  lines.push("    <hudson.tasks.Shell>");
  lines.push(`      <command>${escapeXml(finalCmd)}</command>`);
  lines.push("    </hudson.tasks.Shell>");
  lines.push("  </builders>");
  lines.push("  <publishers>");

  if (email) {
    lines.push(buildEmailPublisherBlock(email, emailCond, emailKeywords, emailRegex, "    "));
  }

  lines.push("  </publishers>");
  lines.push("  <buildWrappers/>");
  lines.push("</project>");

  return lines.join("\n");
}

/**
 * Build a CloudBees Folder config.xml.
 * Mirrors Python build_folder_xml().
 */
export function buildFolderXml(desc = ""): string {
  const lines = [
    "<?xml version='1.1' encoding='UTF-8'?>",
    '<com.cloudbees.hudson.plugins.folder.Folder plugin="cloudbees-folder">',
    `  <description>${escapeXml(desc)}</description>`,
    "  <views/>",
    "  <primaryView>All</primaryView>",
    "  <healthMetrics/>",
    "</com.cloudbees.hudson.plugins.folder.Folder>",
  ];
  return lines.join("\n");
}

/**
 * Build a Pipeline (WorkflowJob) config.xml.
 * Uses CpsFlowDefinition (inline script) with optional publishers and triggers.
 */
export function buildPipelineXml(opts: PipelineXmlOpts): string {
  const {
    desc = "",
    script = "pipeline { agent any; stages { stage('Build') { steps { echo 'hello' } } } }",
    schedule = null,
    email = null,
    emailCond = "failed",
    emailKeywords = null,
    emailRegex = null,
    params = null,
  } = opts;

  const lines: string[] = [
    "<?xml version='1.1' encoding='UTF-8'?>",
    '<flow-definition plugin="workflow-job@2.40">',
    `  <description>${escapeXml(desc)}</description>`,
    "  <keepDependencies>false</keepDependencies>",
    buildParametersProperty(params, "  "),
    `  <definition class="org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition" plugin="workflow-cps@2.90">`,
    `    <script>${escapeXml(script)}</script>`,
    "    <sandbox>true</sandbox>",
    "  </definition>",
    "  <triggers>",
  ];

  const timer = buildTimerTriggerBlock(schedule, "    ");
  if (timer) lines.push(timer);

  lines.push("  </triggers>");
  lines.push("  <publishers>");

  if (email) {
    lines.push(buildEmailPublisherBlock(email, emailCond, emailKeywords, emailRegex, "    "));
  }

  lines.push("  </publishers>");
  lines.push("  <disabled>false</disabled>");
  lines.push("</flow-definition>");

  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// XML patch helpers for updateJobFreestyle (parse + mutate config.xml)
// These use fast-xml-parser for reading and hand-built string replacement
// for writing back, since the update flow re-posts raw XML.
// ---------------------------------------------------------------------------

/** Parse config.xml and extract the presendScript text value. */
function extractPresendScriptFromXml(configXml: string): string | null {
  const doc = xmlParser.parse(configXml) as Record<string, unknown>;
  // Walk: project > publishers > ExtendedEmailPublisher > presendScript
  const project = doc["project"] as Record<string, unknown> | undefined;
  if (!project) return null;
  const publishers = project["publishers"] as Record<string, unknown> | undefined;
  if (!publishers) return null;
  const ext = publishers[
    "hudson.plugins.emailext.ExtendedEmailPublisher"
  ] as Record<string, unknown> | undefined;
  if (!ext) return null;
  const ps = ext["presendScript"];
  return typeof ps === "string" ? ps : null;
}
